### PR TITLE
HTMLMesh: Add support for Image elements.

### PR DESCRIPTION
**Pull request by msub2**
**Description**

This PR is a simple addition that allows images to be included into an HTMLMesh while respecting scaled width/height on the `` element. Below is an example of the following rendered with HTMLMesh (original image was 512x512):
```html
<h1>Test</h1>
<img id="image" src="https://i.imgur.com/jf4lNex.png" width="150" height="150" crossorigin="anonymous">
```

### CODE REVIEW

File: examples/jsm/interactive/HTMLMesh.js [hunk 1]
```diff
@@ -445,6 +445,17 @@ function html2canvas( element ) {
 
 			}
 
+			if ( element instanceof HTMLImageElement ) {
+
+				element.addEventListener( 'load', () => {
```
Comment (Mugen87) [inline examples/jsm/interactive/HTMLMesh.js:450]: Since you are using the `load` event listener: What would happen if the image is already loaded?

Comment (msub2) [reply examples/jsm/interactive/HTMLMesh.js:450]: Ah shoot you're right, I totally forgot about that. It's late right now (probably why I didnt think of it) but I'll fix this in the morning.

File: examples/jsm/interactive/HTMLMesh.js [hunk 2]
```diff
@@ -445,6 +445,27 @@ function html2canvas( element ) {
...
+
+				}
+				
+				if ( element.complete ) {
+
+					drawImage();
+
+				} else {
+					
+					element.addEventListener( 'load', drawImage );
```
[the remark this answers is not in the document]
Comment (msub2) [reply examples/jsm/interactive/HTMLMesh.js:464]: I went ahead and removed it at the start of the drawImage function, similar to how ImageLoader removes them

[the remark this answers is not in the document]
Comment (Mugen87) [reply examples/jsm/interactive/HTMLMesh.js:464]: The event listener is now always removed even in the `element.complete` case. But I believe that does not have a side effect.

File: examples/jsm/interactive/HTMLMesh.js [hunk 3]
```diff
@@ -445,6 +445,28 @@ function html2canvas( element ) {
 
 			}
 
+			if ( element instanceof HTMLImageElement ) {
+
+				const drawImage = () => {
```
Comment (Mugen87) [inline examples/jsm/interactive/HTMLMesh.js:450]: I think it's better to define `drawImage()` outside of `drawElement()`. Otherwise the function gets defined over an over again.

**Review by Mugen87 — DISMISSED** [review]

File: examples/jsm/interactive/HTMLMesh.js [hunk 4]
```diff
@@ -237,6 +237,14 @@ function html2canvas( element ) {
 
 	}
 
+	function drawImage() {
```
Comment (LeviPesin) [inline examples/jsm/interactive/HTMLMesh.js:240]: I think `element`, `x`, `y`, `width`, and `height` variables should be passed here.

Comment (msub2) [reply examples/jsm/interactive/HTMLMesh.js:240]: Yes, I also noticed this not too long ago. I have a fix to submit soon

File: examples/jsm/interactive/HTMLMesh.js [hunk 5]
```diff
@@ -461,7 +461,7 @@ function html2canvas( element ) {
 
 				} else {
 					
-					element.addEventListener( 'load', drawImage( element, x, y, width, height ) );
+					element.addEventListener( 'load', () => drawImage( element, x, y, width, height ) );
```
Comment (LeviPesin) [inline examples/jsm/interactive/HTMLMesh.js:464]: I'm afraid that now callback removal would not work, because `drawImage !== () => drawImage( args )`...

Comment (msub2) [reply examples/jsm/interactive/HTMLMesh.js:464]: You're right, I should have a fix for this as well

**Comment by msub2** [comment]
There seems to be an issue with calling drawImage from the event listener, as I see the listener function firing but am not seeing the actual image being drawn to the canvas. Will need to take some more time to debug

**Comment by Mugen87** [comment]
Closing. The PR is incomplete and can't be merged in its current state.

File: examples/jsm/interactive/HTMLMesh.js (hunk 5 above)
Comment (remmel) [reply examples/jsm/interactive/HTMLMesh.js:464]: Too bad that that PR cannot be merged because of not handling properly when image is not loaded yet. Maybe I'm wrong but I would have not even bother handling that case in the lib, because update `htmlMesh.material.map.update()` can be fired from user code checking when image are loaded or every 1/30s or 1/60s like done with the statgui example https://threejs.org/examples/?q=vr#webxr_vr_sandbox

Comment (remmel) [reply examples/jsm/interactive/HTMLMesh.js:464]: @LeviPesin I don't understand your explanation, for me when a dom element HTMLImageElement will be removed, the call back will be garbage collected automatically

Comment (LeviPesin) [reply examples/jsm/interactive/HTMLMesh.js:464]: I meant that the line `element.removeElementListener( drawImage )` does not work.

Comment (remmel) [reply examples/jsm/interactive/HTMLMesh.js:464]: @LeviPesin you are right, I missed that line

**Comment by LeviPesin** [comment]
Can't we merge this PR in the state of https://github.com/mrdoob/three.js/pull/24043/commits/d7ddf08f1287f9681dd086e2d2bbfff8226487ab?

**Comment by Mugen87** [comment]
see https://github.com/mrdoob/three.js/pull/24043#issuecomment-1125469334

AFAICS, the implementation was not ready for merging yet.

**Comment by LeviPesin** [comment]
I think it was ready in https://github.com/mrdoob/three.js/commit/d7ddf08f1287f9681dd086e2d2bbfff8226487ab but then broke in https://github.com/mrdoob/three.js/pull/24043/commits/596ad4d247af4692133dba98526e9c375822c865.

**Comment by Mugen87** [comment]
It was right to move the function out and not creating it over and over again.

**Comment by LeviPesin** [comment]
But because of this it broke.

**Comment by Mugen87** [comment]
Okay, but still d7ddf08f1287f9681dd086e2d2bbfff8226487ab was not an appropriate candidate for merging.

**Comment by LeviPesin** [comment]
As an idea... I think we can fix this with promises:

```js
function drawImage( element, x, y, width, height ) {

	buildRectPath( x, y, width, height, 2 );
	context.drawImage( element, x, y, element.width, element.height );

}
```

```js
if ( element instanceof HTMLImageElement ) {

	if ( element.complete ) {

		drawImage( element, x, y, width, height );

	} else {

		let resolve;
		const promise = new Promise( res => { resolve = res; } );
		promise.then( () => drawImage( element, x, y, width, height ) ).then( () => element.removeEventListener( 'load', drawImage ) );
		element.addEventListener( 'load', resolve );

	}

}
```

But it still creates a function. And even https://github.com/mrdoob/three.js/pull/24043/commits/5fbcf259e85a67749a6047b463f8ffa9af5b942b does -- it creates a new anonymous function.

I don't think we might somehow _not_ create a callback for each listener... And I'm not sure why this is discouraged.

**Comment by LeviPesin** [comment]
> I don't think we might somehow not create a callback for each listener...

So I think we should revert to using https://github.com/mrdoob/three.js/pull/24043/commits/d7ddf08f1287f9681dd086e2d2bbfff8226487ab.

**Comment by remmel** [comment]
Hi! Because this is not clear if this is the responsability of that code to handle if the image is not loaded or not and  to make it simpler, I would remove the else block. The user code can still listen for onload and force a draw. Keep in mind that in some project this is not even needed because it is re-drawn every seconds or even even 1/30s

**Comment by Mugen87** [comment]
> Because this is not clear if this is the responsability of that code to handle if the image is not loaded or not and to make it simpler, I would remove the else block. 

I like this idea. The code will be indeed much simpler if `HTMLMesh` assumes the image is fully loaded.
